File: WindowsCE/NurSample_vs2008/NurSample/Beeper.cs
using System;
using System.Collections.Generic;
using System.Text;

using NordicId;

namespace NurSample
{
    class Beeper
    {
        // RTTTL example
        //public string KnightRider = "KnightRider:d=4,o=5,b=125:16e,16p,16f,16e,16e,16p,16e,16e,16f,16e,16e,16e,16d#,16e,16e,16e,16e,16p,16f,16e,16e,16p,16f,16e,16f,16e,16e,16e,16d#,16e,16e,16e,16d,16p,16e,16d,16d,16p,16e,16d,16e,16d,16d,16d,16c,16d,16d,16d,16d,16p,16e,16d,16d,16p,16e,16d,16e,16d,16d,16d,16c,16d,16d,16d";

        // Handle of Beeper driver
        static MHLDriver hBeeper = new MHLDriver();
        // Reference counter
        static int ref_counter = 0;
        // Current volume and frequency
        static uint hz = 0;
        static uint volume = 0;

        public Beeper()
        {
            if (ref_counter == 0 && hBeeper != null)
            {
                try
                {
                    hBeeper.Open("Beeper");
                    this.Volume = 10;
                    hz = this.Hz;
                }
                catch (Exception)
                {
                }
            }
            ref_counter++;
        }

        ~Beeper()
        {
            ref_counter--;
            if (ref_counter == 0 && hBeeper != null)
            {
                try
                {
                    hBeeper.Close();
                }
                catch (Exception)
                {
                }
            }
        }

        public uint Volume
        {
            get
            {
                if (!hBeeper.IsOpen())
                    return 0;

                return hBeeper.GetDword("Beeper.Volume");
            }
            set
            {
                if (!hBeeper.IsOpen())
                    return;

                if (value != volume)
                {
                    volume = value;
                    hBeeper.SetDword("Beeper.Volume", value);
                }
            }
        }

        public uint Hz
        {
            get
            {
                if (!hBeeper.IsOpen())
                    return 0;

                return hBeeper.GetDword("Beeper.Hz");
            }
            set
            {
                if (!hBeeper.IsOpen())
                    return;

                if (value != hz)
                {
                    hz = value;
                    hBeeper.SetDword("Beeper.Hz", value);
                }
            }
        }

        public void Beep(uint durration, bool sync)
        {
            if (!hBeeper.IsOpen())
                return;

            if (sync)
                hBeeper.SetDword("Beeper.SyncBeep", durration);
            else
                hBeeper.SetDword("Beeper.Beep", durration);
        }

        public void BeepSeq(string sequence)
        {
            if (!hBeeper.IsOpen())
                return;

            hBeeper.SetString("Beeper.BeepSeq", sequence);
        }

        public void PlayRtttlFile(string fileName, bool sync)
        {
            if (!hBeeper.IsOpen())
                return;

            if (sync)
                hBeeper.SetString("Beeper.RTTTL.PlayFileSync", fileName);
            else
                hBeeper.SetString("Beeper.RTTTL.PlayFile", fileName);
        }

        public void PlayRtttlBuffer(string rttlBuffer, bool sync)
        {
            if (!hBeeper.IsOpen())
                return;

            if (sync)
                hBeeper.SetString("Beeper.RTTTL.PlayBufferSync", rttlBuffer);
            else
                hBeeper.SetString("Beeper.RTTTL.PlayBuffer", rttlBuffer);
        }
    }
}

File: WindowsCE/NurSample_vs2008_VB/NurSampleCE_VB/Beeper.vb
Imports System.Collections.Generic
Imports System.Text

Imports NordicId

Class Beeper
    ' RTTTL example
    'public string KnightRider = "KnightRider:d=4,o=5,b=125:16e,16p,16f,16e,16e,16p,16e,16e,16f,16e,16e,16e,16d#,16e,16e,16e,16e,16p,16f,16e,16e,16p,16f,16e,16f,16e,16e,16e,16d#,16e,16e,16e,16d,16p,16e,16d,16d,16p,16e,16d,16e,16d,16d,16d,16c,16d,16d,16d,16d,16p,16e,16d,16d,16p,16e,16d,16e,16d,16d,16d,16c,16d,16d,16d";

    ' Handle of Beeper driver
    Shared hBeeper As New MHLDriver()
    ' Reference counter
    Shared ref_counter As Integer = 0
    ' Current volume and frequency
    Shared m_hz As UInteger = 0
    Shared m_volume As UInteger = 0

    Public Sub New()
        If ref_counter = 0 AndAlso hBeeper IsNot Nothing Then
            Try
                hBeeper.Open("Beeper")
                Me.Volume = 10
                m_hz = Me.Hz
            Catch generatedExceptionName As Exception
            End Try
        End If
        ref_counter += 1
    End Sub

    Protected Overrides Sub Finalize()
        Try
            ref_counter -= 1
            If ref_counter = 0 AndAlso hBeeper IsNot Nothing Then
                Try
                    hBeeper.Close()
                Catch generatedExceptionName As Exception
                End Try
            End If
        Finally
            MyBase.Finalize()
        End Try
    End Sub

    Public Property Volume() As UInteger
        Get
            If Not hBeeper.IsOpen() Then
                Return 0
            End If

            Return hBeeper.GetDword("Beeper.Volume")
        End Get
        Set(ByVal value As UInteger)
            If Not hBeeper.IsOpen() Then
                Return
            End If

            If Value <> m_volume Then
                m_volume = Value
                hBeeper.SetDword("Beeper.Volume", Value)
            End If
        End Set
    End Property

    Public Property Hz() As UInteger
        Get
            If Not hBeeper.IsOpen() Then
                Return 0
            End If

            Return hBeeper.GetDword("Beeper.Hz")
        End Get
        Set(ByVal value As UInteger)
            If Not hBeeper.IsOpen() Then
                Return
            End If

            If Value <> m_hz Then
                m_hz = Value
                hBeeper.SetDword("Beeper.Hz", Value)
            End If
        End Set
    End Property

    Public Sub Beep(ByVal durration As UInteger, ByVal sync As Boolean)
        If Not hBeeper.IsOpen() Then
            Return
        End If

        If sync Then
            hBeeper.SetDword("Beeper.SyncBeep", durration)
        Else
            hBeeper.SetDword("Beeper.Beep", durration)
        End If
    End Sub

    Public Sub BeepSeq(ByVal sequence As String)
        If Not hBeeper.IsOpen() Then
            Return
        End If

        hBeeper.SetString("Beeper.BeepSeq", sequence)
    End Sub

    Public Sub PlayRtttlFile(ByVal fileName As String, ByVal sync As Boolean)
        If Not hBeeper.IsOpen() Then
            Return
        End If

        If sync Then
            hBeeper.SetString("Beeper.RTTTL.PlayFileSync", fileName)
        Else
            hBeeper.SetString("Beeper.RTTTL.PlayFile", fileName)
        End If
    End Sub

    Public Sub PlayRtttlBuffer(ByVal rttlBuffer As String, ByVal sync As Boolean)
        If Not hBeeper.IsOpen() Then
            Return
        End If

        If sync Then
            hBeeper.SetString("Beeper.RTTTL.PlayBufferSync", rttlBuffer)
        Else
            hBeeper.SetString("Beeper.RTTTL.PlayBuffer", rttlBuffer)
        End If
    End Sub
End Class
